[Quantization] Add a test for QAT + PTQ selective quantization in

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111923
* __->__ #111689

xnnpack quantizer

Summary:
For some workflows you want to quantize some parts of the model via qat
and then continue eager mode training. After training, you want to
export the whole model and perform PTQ on the rest.

Test Plan:
test added

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50510480](https://our.internmc.facebook.com/intern/diff/D50510480)

kimishpatel committed:
[Quantization] Add a test for QAT + PTQ selective quantization in
xnnpack quantizer

Summary:
For some workflows you want to quantize some parts of the model via qat
and then continue eager mode training. After training, you want to
export the whole model and perform PTQ on the rest.

Test Plan:
test added

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50510480).

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e_qat.py:
```diff
@@ -628,3 +630,109 @@ def test_qat_mobilenet_v2(self):
+        for name, child in model.named_children():
+            if isinstance(child, torch.fx.GraphModule):
+                torch.ao.quantization.move_exported_model_to_eval(child)
+                converted_child = convert_pt2e(child)
```
nit: please enable fold_quantize=True, same for the other convert_pt2e

jerryzh168 replied:
please don't forge this one

kimishpatel replied:
oh yeah. I was thinking I am forgetting something. this was it.

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e_qat.py:
```diff
@@ -628,3 +630,109 @@ def test_qat_mobilenet_v2(self):
+        )
+
+        quantizer = XNNPACKQuantizer()
+        quantizer.set_module_type(torch.nn.Linear, None)
```
just checking, does None work already without any changes?

kimishpatel replied:
no it does not

jerryzh168 replied:
sorry, do you mean you need to add the no_module filter or something else?

kimishpatel replied:
yes thats what i mean.

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e_qat.py:
```diff
@@ -628,3 +630,109 @@ def test_qat_mobilenet_v2(self):
+                )
+                quantizer.set_global(quantization_config)
+                traced_child_prepared = prepare_qat_pt2e(traced_child, quantizer)
+                model._modules[name] = traced_child_prepared
```
nit: this one can use `setattr(model, name, ...)` I think, so we don't expose implementation details here

kimishpatel replied:
yeah I thought of doing the same, but this one seems more stable. It is not clear to me if setattr is equivalent of this.

jerryzh168 replied:
I think `setattr` might be more stable/robust: https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/module.py#L1698

kimishpatel replied:
sounds good

kimishpatel commented on torch/ao/quantization/quantizer/xnnpack_quantizer.py:
```diff
@@ -302,7 +317,9 @@ def set_global(self, quantization_config: QuantizationConfig) -> XNNPACKQuantize
         return self
 
     def set_operator_type(
```
not part of this PR so dont want to touch it. 

Separately, I want to figure out if filtering qconfig for module types/names can be or should be done differently.

kimishpatel committed:
Update on "[Quantization] Add a test for QAT + PTQ selective quantization in"


xnnpack quantizer

Summary:
For some workflows you want to quantize some parts of the model via qat
and then continue eager mode training. After training, you want to
export the whole model and perform PTQ on the rest.

Test Plan:
test added

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50510480](https://our.internmc.facebook.com/intern/diff/D50510480)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50510480).

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e_qat.py:
```diff
@@ -628,3 +630,109 @@ def test_qat_mobilenet_v2(self):
+            else:
+                self._prepare_qat_linears(child)
+
+    def _convert_qat_linears(self, model):
```
maybe refactor this to separate the step of 1. move_to_eval 2. convert_pt2e? and 1. can live in 
 pt2e quant utils as well

kimishpatel replied:
I dont still have clarify on how move_to_eval is gonna help with `model.eval` running into errors.

btw this is just a test code, so not sure if refactor is of importance

jerryzh168 replied:
oh my main point is this will be useful for other use cases that want to mix eager + pt2e flow as well, but it can be done later

jerryzh168 replied:
> how move_to_eval is gonna help with model.eval running into errors

the reason is that model.eval() calls `.eval()` on all submodules, and exported submodules only support move_to_eval

kimishpatel committed:
Update on "[Quantization] Add a test for QAT + PTQ selective quantization in"


xnnpack quantizer

Summary:
For some workflows you want to quantize some parts of the model via qat
and then continue eager mode training. After training, you want to
export the whole model and perform PTQ on the rest.

Test Plan:
test added

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50510480](https://our.internmc.facebook.com/intern/diff/D50510480)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50510480).

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e_qat.py:
```diff
@@ -628,3 +630,105 @@ def test_qat_mobilenet_v2(self):
+            else:
+                self._convert_qat_linears(child)
+
+    def test_mixing_qat_ptq(self):
```
oh one thing I'm thinking is that do you want to include a step to save and load qat weights (state_dict) as well?

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e_qat.py:
```diff
@@ -628,3 +630,105 @@ def test_qat_mobilenet_v2(self):
+
+        self._prepare_qat_linears(model)
+
+        after_prepare_result_pt2e = model(*example_inputs)
```
maybe we can do a save here to simulate the case that people save the QAT'ed model

kimishpatel committed:
Update on "[Quantization] Add a test for QAT + PTQ selective quantization in"


xnnpack quantizer

Summary:
For some workflows you want to quantize some parts of the model via qat
and then continue eager mode training. After training, you want to
export the whole model and perform PTQ on the rest.

Test Plan:
test added

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50510480](https://our.internmc.facebook.com/intern/diff/D50510480)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50510480).

kimishpatel committed:
Update on "[Quantization] Add a test for QAT + PTQ selective quantization in"


xnnpack quantizer

Summary:
For some workflows you want to quantize some parts of the model via qat
and then continue eager mode training. After training, you want to
export the whole model and perform PTQ on the rest.

Test Plan:
test added

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50510480](https://our.internmc.facebook.com/intern/diff/D50510480)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50510480).

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50510480).

kimishpatel committed:
Update on "[Quantization] Add a test for QAT + PTQ selective quantization in"


xnnpack quantizer

Summary:
For some workflows you want to quantize some parts of the model via qat
and then continue eager mode training. After training, you want to
export the whole model and perform PTQ on the rest.

Test Plan:
test added

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50510480](https://our.internmc.facebook.com/intern/diff/D50510480)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50510480).
